### PR TITLE
Support for multiple SSH Keys

### PR DESCRIPTION
**Pull request by bicrxm**
This PR is a part of GSoC 2021 Idea: Support Multiple SSH Keys for VMs. 

- Changes are made in the API part so that at the time of Resetting the SSH keys we can give list of keys as input and reset the keys. Also VM deploy can handle multiple keys.
- The UI is changed so that Multiple SSH keys can be passed at the time of deployment and resetting.

### CODE REVIEW

File: server/src/main/java/com/cloud/api/query/QueryManagerImpl.java [hunk 1]
```diff
@@ -992,9 +992,9 @@
             sb.and("affinityGroupId", sb.entity().getAffinityGroupId(), SearchCriteria.Op.EQ);
         }
 
-        if (keyPairName != null) {
-            sb.and("keyPairName", sb.entity().getKeypairName(), SearchCriteria.Op.EQ);
-        }
+        //if (keyPairName != null) {
+        //    sb.and("keyPairName", sb.entity().getKeypairName(), SearchCriteria.Op.EQ);
+        //}
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/api/query/QueryManagerImpl.java:997]: Do we need this commented code?

File: server/src/main/java/com/cloud/api/query/dao/UserVmJoinDaoImpl.java [hunk 2]
```diff
@@ -325,6 +325,10 @@ public UserVmResponse newUserVmResponse(ResponseView view, String objectName, Us
         if (vmDetails != null) {
             Map<String, String> resourceDetails = new HashMap<String, String>();
             for (UserVmDetailVO userVmDetailVO : vmDetails) {
+                if (userVmDetailVO.getName().startsWith(VmDetailConstants.KEY_PAIR_NAMES)) {
+                    s_logger.info(userVmDetailVO.getValue());
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/api/query/dao/UserVmJoinDaoImpl.java:329]: We could improve this log with some context.

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 3]
```diff
@@ -862,17 +862,52 @@ public UserVm resetVMSSHKey(ResetVMSSHKeyCmd cmd) throws ResourceUnavailableExce
             throw new InvalidParameterValueException("Vm " + userVm + " should be stopped to do SSH Key reset");
         }
 
-        SSHKeyPairVO s = _sshKeyPairDao.findByName(owner.getAccountId(), owner.getDomainId(), cmd.getName());
-        if (s == null) {
-            throw new InvalidParameterValueException("A key pair with name '" + cmd.getName() + "' does not exist for account " + owner.getAccountName()
-            + " in specified domain id");
+        if (cmd.getName() == null && cmd.getNames() == null) {
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:865]: `org.apache.commons.lang3.ObjectUtils` has a method to verify nulls `anyNull(...)`.

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 4]
```diff
@@ -862,17 +862,52 @@ public UserVm resetVMSSHKey(ResetVMSSHKeyCmd cmd) throws ResourceUnavailableExce
...
+        if (cmd.getNames() != null) {
+            s_list = _sshKeyPairDao.findByNames(owner.getAccountId(), owner.getDomainId(), cmd.getNames());
+            for (String keypairname : cmd.getNames()) {
+                if (keypairnames != "") {
+                    keypairnames = keypairnames + ", ";
+                }
+                keypairnames = keypairnames + keypairname;
+            }
+        }
+        if (s_list == null && s == null) {
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:886]: `org.apache.commons.lang3.ObjectUtils` has a method to verify nulls `anyNull(...)`.

File: api/src/main/java/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java [hunk 5]
```diff
@@ -153,9 +153,12 @@
     @Parameter(name = ApiConstants.USER_DATA, type = CommandType.STRING, description = "an optional binary data that can be sent to the virtual machine upon a successful deployment. This binary data must be base64 encoded before adding it to the request. Using HTTP GET (via querystring), you can send up to 2KB of data after base64 encoding. Using HTTP POST(via POST body), you can send up to 32K of data after base64 encoding.", length = 32768)
     private String userData;
 
-    @Parameter(name = ApiConstants.SSH_KEYPAIR, type = CommandType.STRING, description = "name of the ssh key pair used to login to the virtual machine")
+    @Parameter(name = ApiConstants.SSH_KEYPAIR, type = CommandType.LIST, collectionType = CommandType.STRING, description = "name of the ssh key pairs used to login to the virtual machine")
```
Comment (davidjumani) [inline api/src/main/java/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java:156]: This should not be changed

File: api/src/main/java/org/apache/cloudstack/api/command/user/vm/ResetVMSSHKeyCmd.java [hunk 6]
```diff
@@ -58,8 +60,11 @@
     @Parameter(name = ApiConstants.ID, type = CommandType.UUID, entityType = UserVmResponse.class, required = true, description = "The ID of the virtual machine")
     private Long id;
 
-    @Parameter(name = ApiConstants.SSH_KEYPAIR, type = CommandType.STRING, required = true, description = "name of the ssh key pair used to login to the virtual machine")
-    private String name;
+    @Parameter(name = ApiConstants.SSH_KEYPAIR, type = CommandType.LIST, collectionType = CommandType.STRING ,description = "name of the ssh key pair to bt used to login to the virtual machine")
```
Comment (davidjumani) [inline api/src/main/java/org/apache/cloudstack/api/command/user/vm/ResetVMSSHKeyCmd.java:63]: This too

File: api/src/main/java/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java [hunk 7]
```diff
@@ -434,8 +437,9 @@ public String getName() {
         return name;
     }
 
-    public String getSSHKeyPairName() {
-        return sshKeyPairName;
+    public List<String> getSSHKeyPairNames() {
+        if(sshKeyPairName != null) sshKeyPairNames.add(sshKeyPairName);
```
Comment (davidjumani) [inline api/src/main/java/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java:441]: Can this cause a null pointer exception if keypairs is not passed ?

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/actionworkers/KubernetesClusterResourceModifierActionWorker.java [hunk 8]
```diff
@@ -146,9 +146,10 @@ private String getKubernetesNodeConfig(final String joinIp, final boolean ejectI
         final String clusterTokenKey = "{{ k8s_control_node.cluster.token }}";
         final String ejectIsoKey = "{{ k8s.eject.iso }}";
         String pubKey = "- \"" + configurationDao.getValue("ssh.publickey") + "\"";
-        String sshKeyPair = kubernetesCluster.getKeyPair();
-        if (!Strings.isNullOrEmpty(sshKeyPair)) {
-            SSHKeyPairVO sshkp = sshKeyPairDao.findByName(owner.getAccountId(), owner.getDomainId(), sshKeyPair);
+        List<String> sshKeyPairs = null;
+        sshKeyPairs.add(kubernetesCluster.getKeyPair());
```
Comment (davidjumani) [inline plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/actionworkers/KubernetesClusterResourceModifierActionWorker.java:150]: Won't this cause a null pointer exception ?

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 9]
```diff
@@ -862,25 +861,48 @@ public UserVm resetVMSSHKey(ResetVMSSHKeyCmd cmd) throws ResourceUnavailableExce
...
+
+        List<SSHKeyPairVO> s_list = null;
+        if (cmd.getNames() != null) {
+            s_list = _sshKeyPairDao.findByNames(owner.getAccountId(), owner.getDomainId(), cmd.getNames());
+            for (String keypairname : cmd.getNames()) {
+                if (keypairnames != "") {
+                    keypairnames = keypairnames + ", ";
+                }
+                keypairnames = keypairnames + keypairname;
+            }
```
Comment (davidjumani) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:874]: Is there a better way to join a list of names ?

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 10]
```diff
@@ -862,25 +861,48 @@ public UserVm resetVMSSHKey(ResetVMSSHKeyCmd cmd) throws ResourceUnavailableExce
...
+                sshPublicKey = sshPublicKey + "\n";
+                s_logger.info("the public key for keypair name " + s_each.getName() + " is " + publicKey);
+            }
+        }
 
+        boolean result = resetVMSSHKeyInternal(vmId, sshPublicKey, keypairnames);
+
+        UserVmVO vm = _vmDao.findById(vmId);
+        _vmDao.loadDetails(vm);
+        _vmDao.saveDetails(vm);
```
Comment (davidjumani) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:898]: Why is it being loaded and then immediately saved ?

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 11]
```diff
@@ -921,12 +942,14 @@ private boolean resetVMSSHKeyInternal(Long vmId, String sshPublicKey) throws Res
...
             _vmDao.loadDetails(userVm);
             userVm.setDetail(VmDetailConstants.SSH_PUBLIC_KEY, sshPublicKey);
+            userVm.setDetail(VmDetailConstants.KEY_PAIR_NAMES, keypairnames);
             _vmDao.saveDetails(userVm);
 
             if (vmInstance.getState() == State.Stopped) {
                 s_logger.debug("Vm " + vmInstance + " is stopped, not rebooting it as a part of SSH Key reset");
                 return true;
             }
+            s_logger.info("bb " + userVm.getDetail(VmDetailConstants.KEY_PAIR_NAMES));
```
Comment (davidjumani) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:952]: Better logging

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 12]
```diff
@@ -3785,17 +3807,32 @@ private UserVm createVirtualMachine(DataCenter zone, ServiceOffering serviceOffe
...
+                if (pair == null) {
+                    throw new InvalidParameterValueException("A key pair with name '" + pair.getName() + "' was not found.");
+                }
+                else {
+                    s_logger.info("publickey is " + pair.getPublicKey());
+                    sshPublicKey = sshPublicKey + pair.getPublicKey();
+                    sshPublicKey = sshPublicKey + "\n";
+                }
+            }
+        }
```
Comment (davidjumani) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:3833]: Can this duplicate code be extracted out ?

File: ui/src/views/compute/ResetSSH.vue [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+        this.$notification.success({
+          message: this.$t('SSH Key(s) resetted sucessfully')
+        })
+        this.$parent.$parent.close()
+      }).catch(error => {
+        this.$notifyError(error)
+      }).finally(() => {
+        this.loading = false
+      })
+      window.location.reload()
```
Comment (davidjumani) [inline ui/src/views/compute/ResetSSH.vue:144]: Try to refresh the data without reloading the entire page

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@bicrxm Please check the build failure

**Comment by bicrxm** [comment]
> @davidjumani a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

**Comment by bicrxm** [comment]
> @davidjumani a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

**Comment by bicrxm** [comment]
> @davidjumani a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

Can you show me the error message?

**Comment by davidjumani** [comment]
@bicrxm 
```
api/src/main/java/org/apache/cloudstack/api/command/user/vm/ResetVMSSHKeyCmd.java:65: Line has trailing spaces. [RegexpSingleline]
```
You can build / test it locally by running `mvn -Psystemvm,developer clean package`

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan test

**Comment by davidjumani** [comment]
@blueorangutan test keepEnv

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 14]
```diff
@@ -862,25 +860,44 @@ public UserVm resetVMSSHKey(ResetVMSSHKeyCmd cmd) throws ResourceUnavailableExce
...
-        SSHKeyPairVO s = _sshKeyPairDao.findByName(owner.getAccountId(), owner.getDomainId(), cmd.getName());
-        if (s == null) {
-            throw new InvalidParameterValueException("A key pair with name '" + cmd.getName() + "' does not exist for account " + owner.getAccountName()
-            + " in specified domain id");
+        String keypairnames = "";
+
+        List<SSHKeyPairVO> s_list = null;
+
+        if (cmd.getNames() != null) {
+            s_list = _sshKeyPairDao.findByNames(owner.getAccountId(), owner.getDomainId(), cmd.getNames());
```
Comment (davidjumani) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:868]: A check to ensure that all the keyparis passed exist

**Comment by davidjumani** [comment]
@bicrxm There appears to be an SQL error, please it and push the changes

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan test keepEnv

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan test keepEnv

**Comment by bicrxm** [comment]
@blueorangutan test keepEnv

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 15]
```diff
@@ -4113,11 +4110,11 @@ public UserVmVO doInTransaction(TransactionStatus status) throws InsufficientCap
                     vm.details.putAll(details);
                 }
 
-                if (sshPublicKey != null) {
+                if (sshPublicKey != "") {
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:4113]: isEmpty() or isBlank() kind of condition, maybe?

**Comment by bicrxm** [comment]
@blueorangutan test

**Comment by bicrxm** [comment]
@blueorangutan  test keepEnv

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan test keepEnv

**Comment by davidjumani** [comment]
@blueorangutan package

File: api/src/main/java/org/apache/cloudstack/api/command/user/vm/ResetVMSSHKeyCmd.java [hunk 16]
```diff
@@ -78,8 +84,11 @@
     /////////////////// Accessors ///////////////////////
     /////////////////////////////////////////////////////
 
-    public String getName() {
-        return name;
+    public List<String> getNames() {
+        List<String> keypairnames = new ArrayList<String>();
+        if(names != null) { keypairnames = names; }
+        if(name != null && !name.isEmpty()) { keypairnames.add(name); }
```
Comment (GutoVeronezi) [inline api/src/main/java/org/apache/cloudstack/api/command/user/vm/ResetVMSSHKeyCmd.java:90]: According to our [Coding conventions](https://cwiki.apache.org/confluence/display/CLOUDSTACK/Coding+conventions), section `Files, Layout and whitespace`, item 7, `if` clauses should be:

```java
if (condition) {
    statements;
}
```

**Review by XDVD271BT — APPROVED** [review]

**Comment by DaanHoogland** [comment]
@bicrxm can you fix the conflicts? It is marked for 4.17, now. Is this otherwise ready for review/merge?

**Comment by nvazquez** [comment]
Hi @bicrxm can you please fix the conflicts? Is this PR ready for review and testing?

**Comment by davidjumani** [comment]
Will be tracked as part of https://github.com/apache/cloudstack/pull/5965

**Pull request closed without merging**